students/HarveyHerela/lesson04/mailroom2.py

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python
+
+import datetime
+import os
+
+donors = [{"firstname": "Chase", "lastname": "Zuma", "donations": [25]},
+          {"firstname": "Marshall", "lastname": "Rocky", "donations": [30, 60]},
+          {"firstname": "Skye", "lastname": "Rubble", "donations": [100, 100, 100]}]
+
+def get_name(name_type):    
+    gettingInput = True
+
+    while gettingInput:
+        
+        print("Type 'list' for a list of names in the database, '~' to go back")
+        if name_type == "first":
+            response = input("First name: ")
+        elif name_type == "last":
+            response = input("Last name: ")
+
+        if response.lower() == "list":
+            for d in donors:
+                print("{firstname} {lastname}".format(**d))
+        else:
+            gettingInput = False
+    return response
+
+def send_thank_you():
+    """Prompts for a donor name and donation amount, then prints a 'thank you' message to the screen."""
+    firstname= get_name("first")
+
+    if firstname == "~":
+        return
+
+    lastname = get_name("last")
+
+    if lastname == "~":
+        return
+
+    print("Donation amount, or '~' to go to main menu:")
+    amount = input(":")
+    if amount == '~':
+        return
+    amount = float(amount)
+    # If it's an existing donor, add it to their donation list.
+    # If it's a new donor, create a new entry.
+    for d in donors:
+        if d["firstname"] == firstname and d["lastname"] == lastname:
+            d["donations"].append(amount)
+            break
+    else:
+        donors.append({"firstname": firstname, "lastname": lastname, "donations": [amount]})
+        
+    thank_you_msg = "Thank you {firstname} {lastname} for your generous donation of ${amount:.2f}".format(
+        firstname=firstname, lastname=lastname, amount=amount)
+    print(thank_you_msg)
+
+def get_total(donor_data):
+    return donor_data[1]
+
+def create_report():
+    """Prints a list of all donors, sorted by total amount given."""
+    report = str()
+    report += "Donor Name    | Total Given   | Num Gifts | Average Gift"
+    report += "\n"
+    report += "--------------------------------------------------------" 
+    report += "\n"
+    # Create a list with the data to be printed
+    data_list = list()
+    for donor_data in donors:
+        fullname = "{firstname} {lastname}".format(**donor_data)
+        gift_data = [fullname, 0, 0, 0]
+        for donation in donor_data["donations"]:
```

Comment:
You could eliminate this inner loop and make your code a bit more readable here by using some of Python's built in functions

```
for donor_data in donors:
        fullname = "{firstname} {lastname}".format(**donor_data)
        total_donations = sum(donor_data["donations"])
        number_of_donations = len(donor_data["donations"])
        avg_donation = total_donations / number_of_donations
        data_list.append(
            [fullname, total_donations, number_of_donations, avg_donation]
        )
```